openmoc/plotter.py

```diff
@@ -515,6 +515,139 @@ def plot_flat_source_regions(geometry, gridsize=250, xlim=None, ylim=None):
+  if not os.path.exists(directory):
+    os.makedirs(directory)
+
+  # Error checking
+  if not 'Geometry' in str(type(geometry)):
+    py_printf('ERROR', 'Unable to plot the centroids since ' + \
+              'input was not a geometry class object')
+
+  if not is_integer(gridsize):
+    py_printf('ERROR', 'Unable to plot the centroids since ' + \
+              'since the gridsize %d is not an integer', gridsize)
+
+  if gridsize <= 0:
+    py_printf('ERROR', 'Unable to plot the centroids ' + \
+              'with a negative gridsize (%d)', gridsize)
+
+  py_printf('NORMAL', 'Plotting the centroids...')
+
+  # Get the number of flat source regions
+  num_fsrs = geometry.getNumFSRs()
+
+  if num_fsrs == 0:
+    py_printf('ERROR', 'Unable to plot the centroids ' + \
+              'since no tracks have been generated.')
+
+  # Initialize a NumPy array for the surface colors and empty list of centroids
+  surface = numpy.zeros((gridsize, gridsize), dtype=np.int64)
+  centroids_x = []
+  centroids_y = []
+
+  # Retrieve the pixel coordinates
+  coords = get_pixel_coords(geometry, gridsize, xlim, ylim)
+
+  # Find the flat source region IDs for each grid point
+  for i in range(gridsize):
+    for j in range(gridsize):
+
+      x = coords['x'][i]
+      y = coords['y'][j]
+
+      local_coords = openmoc.LocalCoords(x, y)
+      local_coords.setUniverse(geometry.getRootUniverse())
+      geometry.findCellContainingCoords(local_coords)
+      fsr_id = geometry.getFSRId(local_coords)
+
+      # If we did not find a region for this region, use a -1 "bad" number color
+      if fsr_id is None:
+        surface[j][i] = -1
+      else:
+       surface[j][i] = fsr_id
+
+      del local_coords
+
+  # Replace each Cell ID with a random (but reproducible) color ID
+  # NOTE: This color coding scheme only works for FSRs and CMFD cells and not
+  # for Materials and Cells. The reason is that FSRs and CMFD cells are by
+  # definition a sequence of consecutive, monotonically increasing integers.
+  # Material and Cell IDs however may be any sequence of positive integers.
+  all_ids = np.arange(num_fsrs, dtype=np.int64)
+
+  id_colors = np.arange(num_fsrs, dtype=np.int64)
+  numpy.random.seed(1)
+  np.random.shuffle(id_colors)
+
+  ids_to_colors = np.arange(num_fsrs, dtype=np.int64)
+  ids_to_colors[all_ids] = id_colors
+
+  colors = ids_to_colors.take(surface)
+
+  # Make Matplotlib color "bad" numbers (ie, NaN, INF) with transparent pixels
+  cmap = plt.get_cmap('spectral')
+  cmap.set_bad(alpha=0.0)
+
+  # Plot a 2D color map of the flat source regions
+  fig = plt.figure()
+  colors = np.flipud(colors)
+  plt.imshow(colors, extent=coords['bounds'],
+             interpolation='nearest', cmap=cmap, vmin=0, vmax=num_fsrs)
+
+  # Plot centroids on top of 2D FSR color map
+  for r in range(geometry.getNumFSRs()):
+    point = geometry.getFSRCentroid(r)
+    x = point.getX()
+    y = point.getY()
+    
+    if xlim is None and ylim is None:
```

Comment:
This conditional is needed in case someone wants to plot only the centroids for a certain region of the geometry. Say someone wants to plot the centroids for one pin cell in an assembly and inputs `xlim` and `ylim` accordingly. If this conditional was not present, the flat source regions would only be plotted for a few the pin cell, but the centroids would be plotted for the entire geometry.

Another option could be to remove the conditional entirely and have the following:

`if xlim is not None:`
`plt.xlim(xlim)`
`if ylim is not None:`
`plt.ylim(ylim)`

What do you think about changing the long condition to just setting the plot limits? The results are the same.